### PR TITLE
add 2nd in series

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 1]
```diff
@@ -0,0 +1,347 @@
+---
+title: "Introducing torch autograd"
+description: > 
+ Instructive though it may be to code backpropagation from scratch, with torch, there is hardly ever a reason to. Its automatic differentiation feature, called autograd, keeps track of operations that need their gradients computed, as well as how to compute them.
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:4]: I think for the description you should have it be as straightforward as possible. So, the first line might read something like: "With torch, there is hardly ever a reason to code backpropagation from scratch…"

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Last week, we saw how to code [a simple network from
+scratch](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+using nothing but `torch` *tensors*. Predictions, loss, gradients,
+weight updates -- all these things we've been computing ourselves.
+Today, we make a significant change: Namely, we spare ourselves the
+cumbersome calculation of gradients, and have `torch` do it for us.
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:34]: I think the list of "cumbersome things" should go in the line where you're saying what you have torch do for you, if it's doing all of those things in this case. If not, I would start the list with gradients so that it's clear that you're knocking off that first item, and not the rest.

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+Prior to that though, let's get some background.
+
+## Automatic differentiation with *autograd*
+
+`Torch` uses a module called *autograd* to
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:40]: Is this torch the package? If so, it should remain lowercase, since R is case-sensitive.

Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:40]:
```suggestion
`torch` uses a module called *autograd* to
```

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+(1) record operations performed on tensors, and
+
+(2) store what will have to be done to obtain the corresponding
+    gradients, once we're entering the backward pass.
+
+These prospective actions are stored internally as functions, and when
+it's time to compute the gradients, these functions are applied in
+order: Application starts from the output node, and calculated gradients
+are successively being *propagated* *back* through the network. This is
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:50]:
```suggestion
are successively *propagated* *back* through the network. This is
```

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+a form of *reverse mode automatic differentiation*.
+
+#### *Autograd* basics
+
+As users, we can see a bit of the implementation. As a prerequisite for
+this "recording" to happen, tensors have to be created with
+`requires_grad = TRUE`. E.g.,
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:57]:
```suggestion
`requires_grad = TRUE`. For example:
```

Comment (batpigandme) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:57]: I just hate seeing E.g. capitalized like that, so I replaced it altogether 😜

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+```{r}
+library(torch)
+
+x <- torch_ones(2, 2, requires_grad = TRUE)
+```
+
+To be clear, `x` now is a tensor *with respect to which* gradients have
+to be calculated -- normally, a tensor representing a weight or a bias,
+not the input data [^1]. If subsequently we perform some operation on
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:67]:
```suggestion
not the input data [^1]. If we subsequently perform some operation on
```

Comment (batpigandme) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:67]: As is, it should be offset with commas, so I just inlined it so that's unnecessary.

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+With longer chains of computations, we can take a glance at how `torch`
+builds up a graph of backward operations. Here is a slightly more
+complex example -- feel free to skip if you're not the type who just
+*has* to peek into things for them to make sense.
+
+#### Digging deeper
+
+We build up a simple graph of tensors, with "inputs" `x1` and `x2` being
+connected to "output" `out` by "intermediaries" `y` and `z`.
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:113]: I'm not sure if you need the scare quotes around inputs, output, etc. There may be some technical reason, though, so totally your call.

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+x2 <- torch_tensor(1.1, requires_grad = TRUE)
+
+y <- x1 * (x2 + 2)
+
+z <- y$pow(2) * 3
+
+out <- z$mean()
+```
+
+To save memory, by default intermediate gradients are not stored.
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:126]:
```suggestion
To save memory, intermediate gradients are not stored by default.
```

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+Calling `retain_grad()` on a tensor allows to deviate from this default;
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:127]: You need something between "allows" and "to" (allows you to?)

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+let's do this here, for the sake of demonstration:
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:128]:
```suggestion
Let's do this here, for the sake of demonstration:
```

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 9 above)
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:127]:
```suggestion
Calling `retain_grad()` on a tensor allows to deviate from this default.
```

Comment (batpigandme) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:127]: With technical writing, I think it's a good thing to simplify, so I sort of "factor out" semi-colon connectors where possible.

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+## The simple network, now using *autograd*
+
+Thanks to *autograd*, we say good-bye to the tedious, error-prone
+process of coding backpropagation ourselves. A single method call does
+it all: `loss$backward()`.
+
+With `torch` keeping track of operations as required, we don't even have
+to explicitly name the intermediate tensors any more. Just for fun and
+to make the point, we can code forward pass, loss calculation and
+backward pass in just three lines:
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:233]: I think this point can be made without the qualifier (especially since you have another "just" in the sentence)…also allows you to use an Oxford comma in the list, which is nice because "pass" is a noun and a verb.

Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:233]:
```suggestion
to explicitly name the intermediate tensors any more. We can code forward pass, loss calculation, and backward pass in just three lines:
```

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+y_pred <- x$mm(w1)$add(b1)$clamp(min = 0)$mm(w2)$add(b2)
+  
+loss <- (y_pred - y)$pow(2)$sum()
+
+loss$backward()
+```
+
+Here is the complete code. We're at an intermediate stage: We still
+manually compute the forward pass and the loss, and we still manually
+update the weights. Due to the latter, there is something I need to
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:245]:
```suggestion
Here is the complete code. We're at an intermediate stage: We still
manually compute the forward pass and the loss, and update the weights. 
Due to the latter, there is something I need to
```

Comment (batpigandme) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:245]: "We still manually" only needs to be written the first time, since it applies to both items in the list. 👍

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+We make use of these fields to update the weights. But now that
+*autograd* is "on", whenever we execute an operation we *don't* want
+recorded for backprop, we need to explicitly exempt it: This is why we
+wrap the weight updates in a call to `with_no_grad()`.
+
+[^2]: All that have `requires_grad` set to `TRUE`, to be precise.
+
+While this is something you may file under "nice to know" -- after all,
+once we arrive at the last post in the series, this manual updating of
+weights will be gone --, the idiom of *zeroing gradients* is here to
```
Comment (batpigandme) [inline _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:332]:
```suggestion
weights will be gone -- the idiom of *zeroing gradients* is here to
```

Comment (batpigandme) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:332]: TL;DR of [this thread](https://english.stackexchange.com/questions/131874/can-i-write-a-comma-followed-by-an-em-dash) (and [this one](https://ell.stackexchange.com/questions/24960/can-a-comma-follow-an-em-dash#:~:text=The%20em%2Ddash%20does%20not,an%20abrupt%20change%20of%20thought), too): since an em dash replaces a comma, you don't put them together.

**Review by batpigandme — APPROVED** [review]
Mainly trying to simplify things for easier reading. 👍

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:40]: Well, I capitalized it because beginning-of-sentence ... but will change if you say so :-)

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 11 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:233]: that is excellent!! I'm being overly explicit sometimes it seems

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 12 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:243]: Thanks -  I get what you're saying, but I think in this case it is ok to stress this some more (perhaps for psychological reasons: assure people that this will not stay as it is now :-))

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 13 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:330]: That's very interesting and good to know! I'm pretty sure that would be different in German (but never googled it, i have to admit)

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:57]: ??? :-)

why do you hate it? Anyways, applying...

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 9 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd]: adding "one"

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 7 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:113]: yeah, i wasn't sure either ... removing ....

File: _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd:4]: yeah, right, updating!

**Comment by skeydan** [comment]
Thanks so much @batpigandme !!

**Comment by skeydan** [comment]
is now https://github.com/rstudio/ai-blog/pull/112
